File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/service/broadcasters/TrackPositionBroadcaster.java
package com.lasthopesoftware.bluewater.client.playback.service.broadcasters;

import android.content.Context;
import android.content.Intent;
import android.support.v4.content.LocalBroadcastManager;

import com.lasthopesoftware.bluewater.client.playback.file.PlayingFile;
import com.lasthopesoftware.bluewater.shared.MagicPropertyBuilder;
import com.namehillsoftware.lazyj.CreateAndHold;
import com.namehillsoftware.lazyj.Lazy;

import org.joda.time.Duration;

import io.reactivex.functions.Consumer;

public class TrackPositionBroadcaster implements Consumer<Duration> {
	private CreateAndHold<LocalBroadcastManager> lazyLocalBroadcastManager;
	private final PlayingFile playingFile;

	public TrackPositionBroadcaster(Context context, PlayingFile playingFile) {
		lazyLocalBroadcastManager = new Lazy<>(() -> LocalBroadcastManager.getInstance(context));
		this.playingFile = playingFile;
	}

	@Override
	public void accept(Duration fileProgress) {
		final Intent trackPositionChangedIntent = new Intent(trackPositionUpdate);
		trackPositionChangedIntent.putExtra(TrackPositionChangedParameters.filePosition, fileProgress.getMillis());
		trackPositionChangedIntent.putExtra(TrackPositionChangedParameters.fileDuration, playingFile.getDuration());

		lazyLocalBroadcastManager.getObject().sendBroadcast(trackPositionChangedIntent);
	}

	private static final MagicPropertyBuilder magicPropertyBuilder = new MagicPropertyBuilder(TrackPositionBroadcaster.class);

	public static final String trackPositionUpdate = magicPropertyBuilder.buildProperty("trackPositionUpdate");

	public static class TrackPositionChangedParameters {
		private static final MagicPropertyBuilder magicPropertyBuilder = new MagicPropertyBuilder(TrackPositionChangedParameters.class);

		public static final String filePosition = magicPropertyBuilder.buildProperty("filePosition");
		public static final String fileDuration = magicPropertyBuilder.buildProperty("fileDuration");
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/service/broadcasters/TrackPositionBroadcaster.kt
package com.lasthopesoftware.bluewater.client.playback.service.broadcasters

import com.lasthopesoftware.bluewater.client.browsing.files.properties.FilePropertyHelpers.duration
import com.lasthopesoftware.bluewater.client.browsing.files.properties.ProvideLibraryFileProperties
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.playback.file.PlayingFile
import com.lasthopesoftware.bluewater.client.playback.file.PositionedProgressedFile
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.messages.TrackPositionUpdate
import com.lasthopesoftware.bluewater.shared.messages.application.SendApplicationMessages
import io.reactivex.rxjava3.functions.Consumer
import org.joda.time.Duration

class TrackPositionBroadcaster(
	private val sendApplicationMessages: SendApplicationMessages,
	private val fileProperties: ProvideLibraryFileProperties
) {

	fun broadcastProgress(libraryId: LibraryId, positionedProgressedFile: PositionedProgressedFile) {
		fileProperties
			.promiseFileProperties(libraryId, positionedProgressedFile.serviceFile)
			.then { p ->
				p.duration?.let { duration ->
					positionedProgressedFile.progress
						.then { progress ->
							sendApplicationMessages.sendMessage(TrackPositionUpdate(progress, duration))
						}
				}
			}
	}

	fun observeUpdates(playingFile: PlayingFile): Consumer<Duration> = TrackPositionObserver(playingFile)

	private inner class TrackPositionObserver(private val playingFile: PlayingFile): Consumer<Duration> {
		override fun accept(fileProgress: Duration) {
			playingFile.duration.then { d ->
				sendApplicationMessages.sendMessage(TrackPositionUpdate(fileProgress, d))
			}
		}
	}
}
